activemodel/lib/active_model/base.rb

```diff
@@ -0,0 +1,413 @@
+
+    class_attribute :validation_exceptions, instance_accessor: false, default: [ValidationError]
+    class_attribute :defines_save_override, instance_accessor: false, default: false
+
+    around_save do |_, block|
+      @persisted = false
+      block.call
+      @persisted = true
+    end
+
+    # New objects can be instantiated as either empty (pass no construction parameter) or pre-set with
+    # attributes but not yet saved (pass a hash with key names matching the associated attribute names).
+    # In both instances, valid attribute keys are determined by the attribute names --
+    # hence you can't have attributes that have not been declared.
+    #
+    # ==== Example
+    #   # Instantiates a single new object with attributes
+    #   User.new(first_name: "Jamie")
+    #
+    #   # Instantiates a single new object with a block
+    #   User.new { |user| user.first_name: "Jamie" }
+    #
+    #   # Instantiates a single new object with attributes and a block
+    #   User.new(first_name: "Jamie") { |user| user.first_name.upcase! }
+    def initialize(attributes = nil, &block)
+      run_callbacks :initialize do
+        super(attributes)
+        tap { yield_self(&block) if block }
+      end
+    end
+
+    # Carry out the model's responsibility. Returns +true+ on success. Raises
+    # ValidationError on failure.
+    #
+    # Override +#save!+ in subclasses:
+    #
+    #   attr_accessor :email, :password, :request
+    #
+    #   def save!
+    #     if (user = User.authenticate_by(email: email, password: password)
+    #       request.cookies[:signed_user_id] = user.signed_user_id
+    #     else
+    #       errors.add(:base, :invalid)
+    #
+    #       raise ActiveModel::ValidationError.new(self)
+    #     end
+    #   end
+    #
+    # ==== Options
+    #
+    # * +:validate+ By default, #save! always runs validations. If any of them fail ActiveModel::ValidationError gets raised, and the model won't be saved. However, if you supply
+    # <tt>validate: false</tt>, validations are bypassed altogether. See Validations for more information.
+    # * +:context+ Forwarded to the call to #validate!. See Validations for more information
+    #
+    # ==== Example
+    #
+    #   session = Session.new(email: "user@example.com", password: "")
+    #   session.save! # => raises ActiveModel::ValidationError
+    #
+    #   # skip validations
+    #   session.save!(validate: true) # => raises ActiveModel::ValidationError
+    #   session.save!(validate: false) # => true
+    #
+    #   # with validation contexts
+    #   session = Session.new(email: "", password: "secret")
+    #   session.save!(context: :sign_in) # => raises ActiveModel::ValidationError
+    #   session.save!(context: :permission_check) # => true
+    def save!(validate: true, context: nil)
+      run_callbacks :save do
+        validate!(context) if validate
+        call
+        true
+      end
+    end
+
+    def call # :nodoc:
+    end
```

Comment:
This got me a bit confused. When defining a model that inherits from `ActiveModel::Base`, should we override `#call` to define the "save" behavior?